### ISSUE TITLE
json.cpp compilation failure due to a gcc bug

### ISSUE THREAD

**Issue by yan12125** [facebook/hhvm#5208]
When I'm compiling hhvm commit cf1b737c4946f00a1b07256dc3244e356851046b, I encounter the following problem:
```
cd /home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly && /usr/bin/c++   -DENABLE_FASTCGI=1 -DENABLE_ZEND_COMPAT=1 -DFOLLY_HAVE_CLOCK_GETTIME=1 -DFOLLY_HAVE_FEATURES_H=1 -DFOLLY_HAVE_MALLOC_H=1 -DFOLLY_HAVE_PTHREAD_ATFORK=1 -DFOLLY_HAVE_PTHREAD_SPINLOCK_T=1 -DFOLLY_HAVE_WEAK_SYMBOLS=1 -DFOLLY_NO_CONFIG=1 -DFOLLY_VERSION=\"0.1\" -DHAVE_BOOST1_49 -DHAVE_CURL_MULTI_WAIT -DHAVE_ELF_GETSHDRSTRNDX -DHAVE_FEATURES_H=1 -DHAVE_LIBDL -DHHVM -DHHVM_DYNAMIC_EXTENSION_DIR=\"/usr/lib/hhvm/extensions\" -DHPHP_OSS=1 -DLIBDWARF_CONST_NAME -DNDEBUG -DNO_LIB_GFLAGS -DNO_TCMALLOC=1 -DPACKAGE=hhvm -DPACKAGE_VERSION=Release -DPHP_MYSQL_UNIX_SOCK_ADDR=\"/run/mysqld/mysqld.sock\" -DRELEASE=1 -DTHRIFT_MUTEX_EMULATE_PTHREAD_TIMEDLOCK -DUSE_EDITLINE -DUSE_JEMALLOC=1 -D_GNU_SOURCE -D_PTHREADS=1 -D_REENTRANT=1 -D__STDC_FORMAT_MACROS -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong --param=ssp-buffer-size=4  -Wall -std=gnu++11 -ffunction-sections -fno-gcse -fno-omit-frame-pointer -Woverloaded-virtual -Wno-deprecated -Wno-strict-aliasing -Wno-write-strings -Wno-invalid-offsetof -fno-operator-names -Wno-error=array-bounds -Wno-error=switch -Werror=format-security -Wno-unused-result -Wno-sign-compare -Wno-attributes -Wno-maybe-uninitialized -Wno-unused-local-typedefs -fno-canonical-system-headers -Wno-deprecated-declarations -Wno-unused-function  -ftrack-macro-expansion=0 -fno-builtin-memcmp -fno-delete-null-pointer-checks -mcrc32 -O3 -DNDEBUG -fno-gcse -momit-leaf-frame-pointer --param max-inline-insns-auto=100 --param early-inlining-insns=200 --param max-early-inliner-iterations=50 -I/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/re2/src -I/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/squangle/src -I/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/webscalesqlclient/src/include -I/usr/include/libxml2 -I/usr/include/double-conversion -I/usr/lib/libzip/include -I/usr/include/libdwarf -I/usr/include/security -I/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/hphp -I/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/fastlz -I/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/timelib -I/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/libafdt/src -I/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/libmbfl -I/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/libmbfl/mbfl -I/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/libmbfl/filter -I/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/mcrouter -I/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly -I/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/thrift -I/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party -I/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm    -o CMakeFiles/folly.dir/folly/json.cpp.o -c /home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/json.cpp
In file included from /home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/FBString.h:59:0,
                 from /home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/dynamic.h:78,
                 from /home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/json.h:44,
                 from /home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/json.cpp:17:
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h: In substitution of ‘template<class RHS, RHS rhs, class LHS> bool folly::detail::greater_than_impl(typename std::enable_if<(rhs < std::numeric_limits<_UIntType>::min()), LHS>::type) [with RHS = long int; RHS rhs = 9223372036854775807l; LHS = bool]’:
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h:447:4:   required from ‘bool folly::greater_than(LHS) [with RHS = long int; RHS rhs = 9223372036854775807l; LHS = bool]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Conv.h:103:5:   required from ‘typename std::enable_if<((std::is_integral<_Tp2>::value && std::is_integral<_Tp>::value) && (! std::is_same< <template-parameter-1-1>, <template-parameter-1-2> >::value)), Tgt>::type folly::to(const Src&) [with Tgt = long int; Src = bool; typename std::enable_if<((std::is_integral<_Tp2>::value && std::is_integral<_Tp>::value) && (! std::is_same< <template-parameter-1-1>, <template-parameter-1-2> >::value)), Tgt>::type = long int]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/dynamic-inl.h:759:30:   required from ‘T folly::dynamic::asImpl() const [with T = long int]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/dynamic-inl.h:392:68:   required from here
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h:407:10: warning: comparison of constant ‘9223372036854775807l’ with boolean expression is always false [-Wbool-compare]
     (rhs < std::numeric_limits<LHS>::min()),
          ^
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h: In substitution of ‘template<class RHS, RHS rhs, class LHS> bool folly::detail::greater_than_impl(typename std::enable_if<(rhs > std::numeric_limits<_UIntType>::max()), LHS>::type) [with RHS = long int; RHS rhs = 9223372036854775807l; LHS = bool]’:
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h:447:4:   required from ‘bool folly::greater_than(LHS) [with RHS = long int; RHS rhs = 9223372036854775807l; LHS = bool]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Conv.h:103:5:   required from ‘typename std::enable_if<((std::is_integral<_Tp2>::value && std::is_integral<_Tp>::value) && (! std::is_same< <template-parameter-1-1>, <template-parameter-1-2> >::value)), Tgt>::type folly::to(const Src&) [with Tgt = long int; Src = bool; typename std::enable_if<((std::is_integral<_Tp2>::value && std::is_integral<_Tp>::value) && (! std::is_same< <template-parameter-1-1>, <template-parameter-1-2> >::value)), Tgt>::type = long int]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/dynamic-inl.h:759:30:   required from ‘T folly::dynamic::asImpl() const [with T = long int]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/dynamic-inl.h:392:68:   required from here
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h:364:10: warning: comparison of constant ‘9223372036854775807l’ with boolean expression is always true [-Wbool-compare]
     (rhs > std::numeric_limits<LHS>::max()),
          ^
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h: In substitution of ‘template<class RHS, RHS rhs, class LHS> bool folly::detail::greater_than_impl(typename std::enable_if<(rhs > std::numeric_limits<_UIntType>::max()), LHS>::type) [with RHS = long int; RHS rhs = 9223372036854775807l; LHS = bool]’:
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h:447:4:   required from ‘bool folly::greater_than(LHS) [with RHS = long int; RHS rhs = 9223372036854775807l; LHS = bool]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Conv.h:103:5:   required from ‘typename std::enable_if<((std::is_integral<_Tp2>::value && std::is_integral<_Tp>::value) && (! std::is_same< <template-parameter-1-1>, <template-parameter-1-2> >::value)), Tgt>::type folly::to(const Src&) [with Tgt = long int; Src = bool; typename std::enable_if<((std::is_integral<_Tp2>::value && std::is_integral<_Tp>::value) && (! std::is_same< <template-parameter-1-1>, <template-parameter-1-2> >::value)), Tgt>::type = long int]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/dynamic-inl.h:759:30:   required from ‘T folly::dynamic::asImpl() const [with T = long int]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/dynamic-inl.h:392:68:   required from here
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h:364:10: warning: comparison of constant ‘9223372036854775807l’ with boolean expression is always true [-Wbool-compare]
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h:364:10: warning: comparison of constant ‘9223372036854775807l’ with boolean expression is always true [-Wbool-compare]
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h: In substitution of ‘template<class RHS, RHS rhs, class LHS> bool folly::detail::greater_than_impl(typename std::enable_if<((rhs <= std::numeric_limits<_UIntType>::max()) && (rhs >= std::numeric_limits<_UIntType>::min())), LHS>::type) [with RHS = long int; RHS rhs = 9223372036854775807l; LHS = bool]’:
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h:447:4:   required from ‘bool folly::greater_than(LHS) [with RHS = long int; RHS rhs = 9223372036854775807l; LHS = bool]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Conv.h:103:5:   required from ‘typename std::enable_if<((std::is_integral<_Tp2>::value && std::is_integral<_Tp>::value) && (! std::is_same< <template-parameter-1-1>, <template-parameter-1-2> >::value)), Tgt>::type folly::to(const Src&) [with Tgt = long int; Src = bool; typename std::enable_if<((std::is_integral<_Tp2>::value && std::is_integral<_Tp>::value) && (! std::is_same< <template-parameter-1-1>, <template-parameter-1-2> >::value)), Tgt>::type = long int]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/dynamic-inl.h:759:30:   required from ‘T folly::dynamic::asImpl() const [with T = long int]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/dynamic-inl.h:392:68:   required from here
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h:386:10: warning: comparison of constant ‘9223372036854775807l’ with boolean expression is always false [-Wbool-compare]
     (rhs <= std::numeric_limits<LHS>::max()
          ^
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h:387:14: warning: comparison of constant ‘9223372036854775807l’ with boolean expression is always true [-Wbool-compare]
       && rhs >= std::numeric_limits<LHS>::min()),
              ^
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h: In substitution of ‘template<class RHS, RHS rhs, class LHS> bool folly::detail::less_than_impl(typename std::enable_if<(rhs <= std::numeric_limits<_UIntType>::min()), LHS>::type) [with RHS = long int; RHS rhs = -9223372036854775808l; LHS = bool]’:
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h:440:4:   required from ‘bool folly::less_than(LHS) [with RHS = long int; RHS rhs = -9223372036854775808l; LHS = bool]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Conv.h:110:5:   required from ‘typename std::enable_if<((std::is_integral<_Tp2>::value && std::is_integral<_Tp>::value) && (! std::is_same< <template-parameter-1-1>, <template-parameter-1-2> >::value)), Tgt>::type folly::to(const Src&) [with Tgt = long int; Src = bool; typename std::enable_if<((std::is_integral<_Tp2>::value && std::is_integral<_Tp>::value) && (! std::is_same< <template-parameter-1-1>, <template-parameter-1-2> >::value)), Tgt>::type = long int]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/dynamic-inl.h:759:30:   required from ‘T folly::dynamic::asImpl() const [with T = long int]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/dynamic-inl.h:392:68:   required from here
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h:374:10: warning: comparison of constant ‘-9223372036854775808l’ with boolean expression is always true [-Wbool-compare]
     (rhs <= std::numeric_limits<LHS>::min()),
          ^
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h: In substitution of ‘template<class RHS, RHS rhs, class LHS> bool folly::detail::less_than_impl(typename std::enable_if<(rhs <= std::numeric_limits<_UIntType>::min()), LHS>::type) [with RHS = long int; RHS rhs = -9223372036854775808l; LHS = bool]’:
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h:440:4:   required from ‘bool folly::less_than(LHS) [with RHS = long int; RHS rhs = -9223372036854775808l; LHS = bool]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Conv.h:110:5:   required from ‘typename std::enable_if<((std::is_integral<_Tp2>::value && std::is_integral<_Tp>::value) && (! std::is_same< <template-parameter-1-1>, <template-parameter-1-2> >::value)), Tgt>::type folly::to(const Src&) [with Tgt = long int; Src = bool; typename std::enable_if<((std::is_integral<_Tp2>::value && std::is_integral<_Tp>::value) && (! std::is_same< <template-parameter-1-1>, <template-parameter-1-2> >::value)), Tgt>::type = long int]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/dynamic-inl.h:759:30:   required from ‘T folly::dynamic::asImpl() const [with T = long int]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/dynamic-inl.h:392:68:   required from here
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h:374:10: warning: comparison of constant ‘-9223372036854775808l’ with boolean expression is always true [-Wbool-compare]
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h:374:10: warning: comparison of constant ‘-9223372036854775808l’ with boolean expression is always true [-Wbool-compare]
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h: In substitution of ‘template<class RHS, RHS rhs, class LHS> bool folly::detail::less_than_impl(typename std::enable_if<(rhs > std::numeric_limits<_UIntType>::max()), LHS>::type) [with RHS = long int; RHS rhs = -9223372036854775808l; LHS = bool]’:
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h:440:4:   required from ‘bool folly::less_than(LHS) [with RHS = long int; RHS rhs = -9223372036854775808l; LHS = bool]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Conv.h:110:5:   required from ‘typename std::enable_if<((std::is_integral<_Tp2>::value && std::is_integral<_Tp>::value) && (! std::is_same< <template-parameter-1-1>, <template-parameter-1-2> >::value)), Tgt>::type folly::to(const Src&) [with Tgt = long int; Src = bool; typename std::enable_if<((std::is_integral<_Tp2>::value && std::is_integral<_Tp>::value) && (! std::is_same< <template-parameter-1-1>, <template-parameter-1-2> >::value)), Tgt>::type = long int]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/dynamic-inl.h:759:30:   required from ‘T folly::dynamic::asImpl() const [with T = long int]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/dynamic-inl.h:392:68:   required from here
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h:364:10: warning: comparison of constant ‘-9223372036854775808l’ with boolean expression is always false [-Wbool-compare]
     (rhs > std::numeric_limits<LHS>::max()),
          ^
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h: In substitution of ‘template<class RHS, RHS rhs, class LHS> bool folly::detail::less_than_impl(typename std::enable_if<((rhs <= std::numeric_limits<_UIntType>::max()) && (rhs > std::numeric_limits<_UIntType>::min())), LHS>::type) [with RHS = long int; RHS rhs = -9223372036854775808l; LHS = bool]’:
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h:440:4:   required from ‘bool folly::less_than(LHS) [with RHS = long int; RHS rhs = -9223372036854775808l; LHS = bool]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Conv.h:110:5:   required from ‘typename std::enable_if<((std::is_integral<_Tp2>::value && std::is_integral<_Tp>::value) && (! std::is_same< <template-parameter-1-1>, <template-parameter-1-2> >::value)), Tgt>::type folly::to(const Src&) [with Tgt = long int; Src = bool; typename std::enable_if<((std::is_integral<_Tp2>::value && std::is_integral<_Tp>::value) && (! std::is_same< <template-parameter-1-1>, <template-parameter-1-2> >::value)), Tgt>::type = long int]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/dynamic-inl.h:759:30:   required from ‘T folly::dynamic::asImpl() const [with T = long int]’
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/dynamic-inl.h:392:68:   required from here
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h:353:10: warning: comparison of constant ‘-9223372036854775808l’ with boolean expression is always true [-Wbool-compare]
     (rhs <= std::numeric_limits<LHS>::max()
          ^
/home/yen/Computer/Programming/aur/hhvm-git/src/hhvm/third-party/folly/folly/Traits.h:354:14: warning: comparison of constant ‘-9223372036854775808l’ with boolean expression is always false [-Wbool-compare]
       && rhs > std::numeric_limits<LHS>::min()),
              ^
‘
Internal compiler error: Error reporting routines re-entered.
Please submit a full bug report,
with preprocessed source if appropriate.
See <https://bugs.archlinux.org/> for instructions.
third-party/folly/CMakeFiles/folly.dir/build.make:537: recipe for target 'third-party/folly/CMakeFiles/folly.dir/folly/json.cpp.o' failed
```
This is a gcc bug. I've reported the same issue to https://bugs.archlinux.org/task/44734 and https://gcc.gnu.org/bugzilla/show_bug.cgi?id=65882. Seems the flag ```-Wbool-compare```, which is added in gcc 5.1.0, causes the problem. I add ```CXXFLAGS=-Wno-bool-compare``` when invoking cmake and json.cpp can be compiled with errors. I think it should be a good practice to fix these warnings, or add ```-Wno-bool-compare``` as a temporary workaround. Any ideas?

Environment:
Arch Linux x86_64
GCC: 5.1.0

**Comment by yan12125** [comment]
By the way, I didn't encounter any problem in compiling folly alone because folly does not use ```-Wall``` in ```$CXXFLAGS```

**Comment by jwatzman** [comment]
See also #5184 -- we want to eventually compile with GCC 5, but aren't quite there yet. (And thanks for reporting the GCC bug, so we can either work around, or require a version where it's eventually fixed.)

**Comment by yan12125** [comment]
With gcc 5.1.0-4 on Arch Linux (with snapshot 5-20150519 ftp://gcc.gnu.org/pub/gcc/snapshots/5-20150519/), the problem does not exist anymore. Hope the information helps.

**Comment by yan12125** [comment]
According to https://gcc.gnu.org/bugzilla/show_bug.cgi?id=65882:

> Fixed for GCC 5.2 and 6.

In the latest release of GCC it's fixed. I think this ticket can be closed?
